sympy/combinatorics/fp_groups.py

```diff
@@ -0,0 +1,724 @@
+            i += 1
+        # can this be replaced with i == r ?
+        if i > j:
+            if f != b:
+                # implement the "coincidence" routine on Pg 158 of Handbook.
+                self.coincidence(f, b)
+            return
+        while j >= i and self.table[b][A_dict_inv[word[j]]] is not None:
+            b = self.table[b][A_dict_inv[word[j]]]
+            j -= 1
+        if j < i:
+            # we have an incorrect completed scan with coincidence f ~ b
+            # run the "coincidence" routine
+            self.coincidence(f, b)
+        elif j == i:
+            # deduction process
+            self.table[f][A_dict[word[i]]] = b
+            self.table[b][A_dict_inv[word[i]]] = f
+        # otherwise scan is incomplete and yields no information
+
+    def merge(self, k, lamda, q):
+        p = self.p
+        phi = self.rep(k)
+        chi = self.rep(lamda)
+        if phi != chi:
+            mu = min(phi, chi)
+            v = max(phi, chi)
+            p[v] = mu
+            q.append(v)
+
+    def rep(self, k):
+        p = self.p
+        lamda = k
+        rho = p[lamda]
+        while rho != lamda:
+            lamda = rho
+            rho = p[lamda]
+        mu = k
+        rho = p[mu]
+        while rho != lamda:
+            p[mu] = lamda
+            mu = rho
+            rho = p[mu]
+        return lamda
+
+    # alpha, beta coincide
+    # here alpha, beta represent the pair of cosets where
+    # coincidence occurs
+    def coincidence(self, alpha, beta):
+        """
+        """
+        A_dict = self.A_dict
+        A_dict_inv = self.A_dict_inv
+        p = self.p
+        l = 0
+        # behaves as a queue
+        q = []
+        self.merge(alpha, beta, q)
+        while len(q) > 0:
+            gamma = q.pop(0)
+            # comment by Kalevi, this is already done by p[v] = mu
+            # del C[gamma]
+            # commenting this out
+            # omega = omega - set(gamma)
+            for x in A_dict:
+                delta = self.table[gamma][A_dict[x]]
+                if delta is not None:
+                    self.table[delta][A_dict_inv[x]] = None
+                    mu = self.rep(gamma)
+                    nu = self.rep(delta)
+                    if self.table[mu][A_dict[x]] is not None:
+                        self.merge(nu, self.table[mu][A_dict[x]], q)
+                    elif self.table[nu][A_dict_inv[x]] is not None:
+                        self.merge(mu, self.table[nu][A_dict_inv[x]], q)
+                    else:
+                        self.table[mu][A_dict[x]] = nu
+                        self.table[nu][A_dict_inv[x]] = mu
+
+    # method used in the HLT strategy
+    def scan_and_fill(self, alpha, word):
+        A_dict = self.A_dict
+        A_dict_inv = self.A_dict_inv
+        r = len(word)
+        l_A = len(A_dict)
+        f = alpha
+        i = 0
+        i_A = 0
+        b = alpha
+        j = r - 1
+        while i_A < l_A:
```

Comment:
This could probably be `while True:`. The loop is expected to run until `i > j`.